### PR TITLE
Imporve CTE local coordinate

### PR DESCRIPTION
**Pull request by Seaven**
before:
* MultiCastFragment will send data to all instance 

after
* MultiCastFragment will send data to itself instance
 
In fact, the output fragment of MultiCastFragment always keep same with MultiCastFragment, and they don't need to shuffle data in a specific way. So the best way is just send to itself direct, it's can be avoid data hash compute and network

Other, add CTE produce complexity score, avoid some simplier plan(one join/one agg) reuse cte direct

### CODE REVIEW

**Comment by Seaven** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 1]
```diff
@@ -312,6 +313,9 @@
     @VarAttr(name = CBO_CTE_REUSE)
     private boolean cboCteReuse = false;
 
+    @VarAttr(name = CBO_CTE_REUSE_RATE)
+    private double cboCTERuseRatio = 2.0;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:317]: make it invisible

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEUtils.java [hunk 2]
```diff
@@ -2,14 +2,26 @@
...
 import com.starrocks.sql.optimizer.operator.logical.LogicalCTEProduceOperator;
+import com.starrocks.sql.optimizer.statistics.ColumnStatistic;
 import com.starrocks.sql.optimizer.statistics.Statistics;
 import com.starrocks.sql.optimizer.statistics.StatisticsCalculator;
 
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
 public class CTEUtils {
+    private static final Map<OperatorType, Integer> OPERATOR_SCORES = ImmutableMap.<OperatorType, Integer>builder()
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEUtils.java:18]: should comment what's the meaning of score and how to set the operator score.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEUtils.java [hunk 3]
```diff
@@ -55,6 +68,34 @@ public static void collectCteOperatorsWithoutCosts(Memo memo, OptimizerContext c
         collectCteConsume(memo.getRootGroup(), context, false);
     }
 
+    private static void collectCteProduceEstimate(OptimizerContext context) {
+        CTEContext cteContext = context.getCteContext();
+        for (Map.Entry<Integer, OptExpression> entry : cteContext.getAllCTEProduce().entrySet()) {
+            AtomicInteger scores = new AtomicInteger(0);
+            cteProduceEstimate(entry.getValue().getGroupExpression().getGroup(), scores);
+
+            // score 15 = 1 join + 2 scan + 3 project = 1 agg + 1 scan + 2 project = (1 scan + 1 project) * 3
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEUtils.java:77]: confusing comment

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run be_ut_test_wangrui

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEUtils.java [hunk 4]
```diff
@@ -131,6 +177,14 @@ private static void calculateStatistics(OptExpression expr, OptimizerContext con
         StatisticsCalculator statisticsCalculator = new StatisticsCalculator(
                 expressionContext, context.getColumnRefFactory(), context);
         statisticsCalculator.estimatorStats();
-        expr.setStatistics(expressionContext.getStatistics());
+
+        if (OperatorType.LOGICAL_OLAP_SCAN.equals(expr.getOp().getOpType()) &&
+                expressionContext.getStatistics().getColumnStatistics().values().stream()
+                        .anyMatch(ColumnStatistic::isUnknown)) {
+            // Mark output rows is zero, inline CTE
+            expr.setStatistics(Statistics.buildFrom(expressionContext.getStatistics()).setOutputRowCount(0).build());
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEUtils.java:185]: plz add some comment to describe why need specail treat

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 5]
```diff
@@ -1903,7 +1910,7 @@ public FragmentExecParams(PlanFragment fragment) {
             this.fragment = fragment;
         }
 
-        List<TExecPlanFragmentParams> toThrift(int backendNum, boolean isEnablePipelineEngine) {
+        List<TExecPlanFragmentParams> toThrift(int backendNum, boolean isEnablePipelineEngine) throws Exception {
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:1913]: Get machine maybe throw IOException

**Review by Youngwb — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Pull request merged**